### PR TITLE
Fix cursor position on monaco auto save

### PR DESCRIPTION
**Pull request by habdelra**
This was really tricky because the module contents resource runs on every module file save and as a result the getters in code mode compute a new cursor position on each save. this comes down to realizing that changes the monaco modifier receives that coincide with modifications within the monaco editor are ultimately just SSE server echoes of auto-saved monaco content. I added an SSE echo debounce to the monaco modifier and that solves this problem.

monaco-cursor

### CODE REVIEW

**Review by lukemelia — COMMENTED** [review]
Any way to cover this with a test?

**Comment by habdelra** [comment]
> Any way to cover this with a test?

I’ll give it a shot. My fear I that Monaco assertions seem to be a bit flaky in general, but let’s see…

**Comment by FadhlanR** [comment]
Why not prevent content updates while the user is editing? I did this [3c4303c](https://github.com/cardstack/boxel/pull/792/commits/3c4303c3917a31230725aea8e5dbacda99fb22d0) in my PR to fix this problem.


https://github.com/cardstack/boxel/assets/12637010/3de1e389-8ac4-40df-a613-cd01ae166201

**Comment by habdelra** [comment]
Actually i think i found the source of a lot of monaco flakiness in our tests---there is a 500ms debounce in monaco modifier--but if the component that includes the monaco modifier is destroyed before 500ms of content being written into monaco the content will never appear in monaco. we need to add a "flush" to our monaco modifier to flush all the content out before running test assertions (and probably before consuming component destruction)

**Comment by habdelra** [comment]
> Why not prevent content updates while the user is editing? I did this [3c4303c](https://github.com/cardstack/boxel/pull/792/commits/3c4303c3917a31230725aea8e5dbacda99fb22d0) in my PR to fix this problem.
> 
>  Screen.Recording.2023-11-07.at.11.52.30.mov

i think this approach is much simpler. note how little the monaco modifier has had to change and how little state we are using to determine if we should update the monaco model

**Comment by habdelra** [comment]
actually i spotted a missing `await` in the monaco modifier. likely this is playing a part in monaco flakiness in tests

**Comment by habdelra** [comment]
> Why not prevent content updates while the user is editing? I did this [3c4303c](https://github.com/cardstack/boxel/pull/792/commits/3c4303c3917a31230725aea8e5dbacda99fb22d0) in my PR to fix this problem.
> 
>  Screen.Recording.2023-11-07.at.11.52.30.mov

@FadhlanR I took a deeper look at your change--apologies I never got a chance to review it before it was merged. you made an incorrect assumption around how the monaco model is updated, and your PR actually introduces a bug that will prevent the monaco model from being updated for out of band index changes (SSE events)

**Comment by habdelra** [comment]
@FadhlanR I've reconciled your changes with my changes--wanna take a look?. i've kept the just about all your stuff, but I'm using the last modified time as the heuristic for whether or not the monaco editor is being edited. this eliminates the SSE echo for auto save changes of the monaco editor. all other SSE events that fall outside of this heuristic are valid updates that monaco should reflect in the editor (e.g. like you change the file directly using VS code in your local env).

**Comment by habdelra** [comment]
> Actually i think i found the source of a lot of monaco flakiness in our tests---there is a 500ms debounce in monaco modifier--but if the component that includes the monaco modifier is destroyed before 500ms of content being written into monaco the content will never appear in monaco. we need to add a "flush" to our monaco modifier to flush all the content out before running test assertions (and probably before consuming component destruction)

instead of a flush, it looks a lot easier to just set this debounce to 0ms in the test env. in terms of component destruction, i forgot there is already a mechanism that is taking care of auto save on code mode close--so we should be good there

File: packages/host/app/modifiers/monaco.ts [hunk 1]
```diff
@@ -109,7 +111,8 @@ export default class Monaco extends Modifier<Signature> {
 
   private onContentChanged = restartableTask(
     async (contentChanged: (text: string) => void) => {
-      timeout(DEBOUNCE_MS);
+      this.lastModified = Date.now();
+      await timeout(monacoDebounceMs);
```
Comment (habdelra) [inline packages/host/app/modifiers/monaco.ts:115]: found a missing await!

**Pull request merged**